### PR TITLE
Optimize Az Mcp CLI Startup Time

### PR DESCRIPTION
**Pull request by g2vinay**
Reduces CLI startup overhead by **~39%** (945 ms → 573 ms) for targeted commands like `azmcp storage account get` by skipping initialization of unrelated service areas.

## Changes

All changes are in `Program.cs`:

- **`GetTargetAreaName()`** — detects the leading area token and validates it against the registered areas array; unknown tokens (typos) fall back to full init, preserving "Did you mean?" UX.
- **`ConfigureServices()`** — skips DI registration for the 60+ non-matching `AzureServices` areas when a target area is known; server-start and plugin-telemetry paths always use full init.
- **Concurrent telemetry init** — `InitializeServicesAsync` runs concurrently with `CommandFactory` resolution so neither blocks the other on the hot path.
- **`--version` / `-v` fast path** — handled before DI container construction entirely.
- **`Array.Exists()`** replaces `LINQ .Any()` on the hot path — eliminates a delegate + enumerator heap allocation per invocation.

## Performance impact

| Metric | Before | After | Delta |
|---|---|---|---|
| Startup overhead (`--learn`) | ~1 100 ms | ~750 ms | −32% |
| Startup overhead (real command) | 945 ms | 573 ms | **−39%** |

### CODE REVIEW

File: core/Microsoft.Mcp.Core/src/Commands/CommandFactoryOptions.cs
Comment (alzimmermsft) [inline core/Microsoft.Mcp.Core/src/Commands/CommandFactoryOptions.cs:1]: Is adding this necessary given we perform filtering in `Program.cs`?

Comment (g2vinay) [reply core/Microsoft.Mcp.Core/src/Commands/CommandFactoryOptions.cs:1]: yeah, not required to be added, logic can be controlled via Program cs only.

**Review by jongio — COMMENTED** [review]
Clean optimization. The area-filtering approach is sound - infrastructure areas always register, only AzureServices areas get skipped, and unknown tokens fall back to full init. The null stub pattern already exists in Fabric/Template servers so this is consistent.

One gap: `GetTargetAreaName` has 4 distinct branches (empty/option-prefixed args, "tools" special case, unknown token fallback, happy path) but no dedicated tests. The integration tests cover `ConfigureServices` with a filter, but a regression in the area detection logic (e.g., someone adds a global option that precedes the area token) wouldn't be caught. Consider adding a parameterized test through the public surface that exercises each branch.

The concurrent telemetry init pattern is fine for a CLI tool - the comment explaining the tradeoff is clear and the edge case (DI failure + simultaneous telemetry failure) is genuinely acceptable here.

File: servers/Azure.Mcp.Server/src/Program.cs [hunk 1]
```diff
@@ -301,14 +353,48 @@ internal static async Task InitializeServicesAsync(IServiceProvider serviceProvi
...
+    /// </summary>
+    /// <param name="args">Command-line arguments.</param>
+    /// <returns>The area name (e.g. "storage"), or <see langword="null"/>.</returns>
+    private static string? GetTargetAreaName(string[] args)
+    {
+        var firstToken = args.Length > 0 ? args[0] : null;
+        if (firstToken is not { Length: > 0 } || firstToken.StartsWith('-'))
+        {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply servers/Azure.Mcp.Server/src/Program.cs]: This is true and we might need a generic CLI parsing done first and take the first subcommand:

https://learn.microsoft.com/dotnet/standard/commandline/syntax#tokens

[the remark this answers is not in the document]
Comment (g2vinay) [reply servers/Azure.Mcp.Server/src/Program.cs]: Fixed in the current implementation, GetTargetAreaName now scans all args with a foreach, skipping any token starting with -, and stops at -- (POSIX end-of-options marker). So --debug storage correctly resolves to "storage". Covered by the new parameterized tests in GetTargetAreaNameTests (e.g. ["--namespace", "storage"] → "storage").

File: servers/Azure.Mcp.Server/src/Program.cs [hunk 2]
```diff
@@ -61,15 +67,25 @@ private static async Task<int> Main(string[] args)
...
+            // from the thread pool. CommandFactory resolves command singletons for the target area.
+            // Both complete in parallel so neither adds to the other's latency on the hot path.
+            //
+            // Error handling: if GetRequiredService throws, telemetryInitTask is abandoned rather than
+            // awaited. In .NET the unobserved exception is a diagnostic warning, not a process crash.
+            // That edge case (fatal DI failure + simultaneous telemetry failure) is acceptable because
+            // the DI failure is already reported to the user via the outer catch block.
+            var telemetryInitTask = InitializeServicesAsync(serviceProvider);
             var commandFactory = serviceProvider.GetRequiredService<ICommandFactory>();
+            await telemetryInitTask;
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply servers/Azure.Mcp.Server/src/Program.cs:96]: added the logic to observe exception.

File: servers/Azure.Mcp.Server/src/Program.cs [hunk 3]
```diff
@@ -259,23 +286,48 @@ internal static void ConfigureServices(IServiceCollection services)
...
+            }
             services.AddSingleton(area);
             area.ConfigureServices(services);
         }
 
-        services.AddRegistryRoot(thisAssembly, $"registry.json");
+        // Optimization: server-mode providers (registry, instructions, plugin allowlists) are only
+        // used when running as an MCP server. For CLI area invocations they are never resolved, so
+        // register lightweight stubs to avoid reading embedded resources on every CLI call.
+        if (areaFilter == null || string.Equals(areaFilter, "server", StringComparison.OrdinalIgnoreCase))
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply servers/Azure.Mcp.Server/src/Program.cs]: Fixed via ServerAreaName, it is now a static field derived directly from the registered ServerSetupinstance at startup.

**Review by alzimmermsft — APPROVED** [review]

**Review by jongio — COMMENTED** [review]
Addresses all prior feedback cleanly:

- `GetTargetAreaName` now scans for the first non-option token instead of blindly taking `args[0]`. Handles `--` correctly as end-of-options marker. The acknowledged trade-off (e.g. `--subscription sub1 storage` falls back to full init because `sub1` is picked up first) is safe since full init is always the correct fallback.
- Unobserved task on the telemetry path is now explicitly observed via `SuppressThrowing` in the catch. Clean pattern.
- The `"server"` magic string is derived from the registered `ServerSetup` instance.
- Parameterized tests cover all 4 branches of `GetTargetAreaName` with good edge cases.

No new issues. Ship it.
